### PR TITLE
[GLib] Most public types should be final; most class structs should be private

### PR DESCRIPTION
**Pull request by aperezdc**
#### 234799572ddb4c3f4de5bce7784441165474d038
<pre>
[GLib] Most public types should be final; most class structs should be private
<a href="https://bugs.webkit.org/show_bug.cgi?id=243663">https://bugs.webkit.org/show_bug.cgi?id=243663</a>

Reviewed by Michael Catanzaro.

Switch over to using the WEBKIT}_DECLARE_{DERIVABLE,FINAL}_TYPE in
headers for the 2022 API, and provide a version of the macros which
expands to the pre-2022 code when the new API is not in use. By always
using macros to generate boilerplate there is less manually-written code
to maintain, which is always a good thing. There are some caveats:

- For the old API the WEBKIT_*_FOO macros have to be kept, because the
  macros that generate boilerplate produce those as static inline
  functions and existing code might be relying on them being macros.
  This results using guards in the code, which is not great, but there
  is not a better option.

- The G_DECLARE_DERIVABLE_TYPE macro produces itself the instance struct,
  with the parent type struct as only member. Using it would need
  modifying the code to avoid using the &quot;priv&quot; member, which would be
  some churn. Moreover, the &quot;priv&quot; member is used by the for the
  old API and making every access to it conditional would be
  cumbersome. A modified version of the macro that keeps the &quot;priv&quot;
  member is used instead.

  While in theory we could save 8 bytes (or 4 in 32-bit) per instance
  using G_ADD_PRIVATE, in practice GLib rounds up the size of instance
  to ensure its beginning is aligned, which in cases would practically
  cancel out the savings. Also, API objects are not so numerous, and
  therefore the potential memory savings are a moot point.

  Tackling the removal of the &quot;priv&quot; instance members might be done
  at some point in the future.

- On the other hand, while G_DECLARE_FINAL_TYPE also creates the class
  struct without any frills, we can use the macro because it is allowed
  to turn later on a final type into a non-final one without breaking
  the ABI. Hence, this GLib macro gets reused with a few additions.

While at it, derivable types (WebKitWebView, WebKitInputMethodContext,
WebKitPolicyDecision) have been changed to use the DERIVABLE version of
the macros. This needs to be done in the same patch because the
subclasses of WebKitPolicyDecision are final and the generated
boilerplace includes definitions for autocleanups, and expects that the
parent type has their autocleanups defined in a certain way. Conversely,
the WebKitAutocleanups.h.in header needed a touch-up to avoid providing
duplicate definitions of the autocleanups generated by the boilerplate
macros.

The instance struct for the types which are now final has been moved
into the implementation files, by piggybacking into the
WEBKIT_DEFINE_FINAL_TYPE[_WITH_CODE] macros to emit its definition.

Finally, the _IN_2022_API suffix gets removed from the
WEBKIT_DEFINE_FINAL_TYPE_WITH_CODE_IN_2022_API and
WEBKIT_DEFINE_FINAL_TYPE_IN_2022_API for the sake of brevity, and a
few more types are marked as final.

* Source/WTF/wtf/glib/WTFGType.h:
* Source/WebKit/PlatformGTK.cmake: Install WebKitWebExtensionAutocleanups.h
  only for the old API.
* Source/WebKit/UIProcess/API/glib/WebKitAuthenticationRequest.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitAuthenticationRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitAutocleanups.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitAutomationSession.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitAutomationSession.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitBackForwardList.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitBackForwardList.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitBackForwardListItem.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitBackForwardListItem.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitContextMenu.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitContextMenuItem.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitCookieManager.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitDefines.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitDeviceInfoPermissionRequest.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitDeviceInfoPermissionRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitDownload.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitEditorState.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitEditorState.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitFileChooserRequest.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitFileChooserRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitFindController.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitFindController.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitFormSubmissionRequest.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitFormSubmissionRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitGeolocationManager.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitGeolocationManager.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitGeolocationPermissionRequest.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitGeolocationPermissionRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitHitTestResult.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitInputMethodContext.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitMediaKeySystemPermissionRequest.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitMediaKeySystemPermissionRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitNavigationPolicyDecision.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitNavigationPolicyDecision.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitNotification.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitNotification.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitNotificationPermissionRequest.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitNotificationPermissionRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitOptionMenu.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitOptionMenu.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitPointerLockPermissionRequest.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitPolicyDecision.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitResponsePolicyDecision.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitResponsePolicyDecision.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitSecurityManager.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitSecurityManager.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitURIRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitURIResponse.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitURISchemeRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitURISchemeResponse.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitURISchemeResponse.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitUserContentFilterStore.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitUserContentFilterStore.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitUserMediaPermissionRequest.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitUserMediaPermissionRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitUserMessage.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebContext.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebResource.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebResource.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataAccessPermissionRequest.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataAccessPermissionRequest.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebsitePolicies.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebsitePolicies.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWindowProperties.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWindowProperties.h.in:
* Source/WebKit/UIProcess/API/gtk/WebKitColorChooserRequest.cpp:
* Source/WebKit/UIProcess/API/gtk/WebKitColorChooserRequest.h.in:
* Source/WebKit/UIProcess/API/gtk/WebKitPointerLockPermissionRequest.h.in:
* Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp:
* Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.h.in:
* Source/WebKit/UIProcess/API/gtk/WebKitWebInspector.cpp:
* Source/WebKit/UIProcess/API/gtk/WebKitWebInspector.h.in:
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.h.in:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitFrame.cpp:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitFrame.h.in:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitScriptWorld.cpp:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitScriptWorld.h.in:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebEditor.cpp:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebEditor.h.in:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebExtension.cpp:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebExtension.h.in:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.cpp:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.h.in:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.h.in:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/webkit-web-extension.h.in:

Canonical link: <a href="https://commits.webkit.org/260190@main">https://commits.webkit.org/260190@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/629e8e09b1d63a9e0eb749cec04de3eaf2743734

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107487 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16549 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40390 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116650 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116071 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17974 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7791 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99651 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113248 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13573 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96758 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41196 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95478 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28384 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82967 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96824 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9555 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29737 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/96246 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7554 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10212 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6631 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15711 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49318 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105094 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7048 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11772 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26061 "Passed tests") | 
| | | | |

### CODE REVIEW

File: Source/JavaScriptCore/API/glib/JSCContextInternal.h [hunk 1]
```diff
@@ -19,6 +19,6 @@
 
 #pragma once
 
-#include "JSCContext.h"
+#include "jsc/JSCContext.h"
```
Comment (carlosgcampos) [inline Source/JavaScriptCore/API/glib/JSCContextInternal.h]: Why do we need this change?

File: Source/JavaScriptCore/API/glib/JSCContextPrivate.h [hunk 2]
```diff
@@ -19,8 +19,6 @@
 
 #pragma once
 
-#include "JSCContext.h"
-#include "JSCValue.h"
```
Comment (carlosgcampos) [inline Source/JavaScriptCore/API/glib/JSCContextPrivate.h]: And this? I think it's ok to include instead of forward declare in these private headers.

File: Source/JavaScriptCore/API/glib/JSCDefines.h.in [hunk 3]
```diff
@@ -47,4 +47,30 @@
...
+#define JSC_DECLARE_FINAL_TYPE(ModuleObjName, module_obj_name, MODULE, OBJ_NAME, ParentName) \
+    JSC_API G_DECLARE_FINAL_TYPE(ModuleObjName, module_obj_name, MODULE, OBJ_NAME, ParentName) \
+                                                                           \
+    typedef struct _ ## ModuleObjName ## Private ModuleObjName ## Private; \
+                                                                           \
+    struct _ ## ModuleObjName {                                            \
+        ParentName parent;                                                 \
+        /*< private >*/                                                    \
+        ModuleObjName ## Private *priv;                                    \
+    };
```
Comment (carlosgcampos) [inline Source/JavaScriptCore/API/glib/JSCDefines.h.in]: Can't this be moved to the cpp?

File: Source/JavaScriptCore/API/glib/JSCDefines.h.in [hunk 4]
```diff
@@ -47,4 +47,30 @@
...
+                                                                           \
+    typedef struct _ ## ModuleObjName ModuleObjName;                       \
+    typedef struct _ ## ModuleObjName ## Class ModuleObjName ## Class;     \
+    typedef struct _ ## ModuleObjName ## Private ModuleObjName ## Private; \
+                                                                           \
+    struct _ ## ModuleObjName {                                            \
+        ParentName parent;                                                 \
+        /*< private >*/                                                    \
+        ModuleObjName ## Private *priv;                                    \
+    };
```
Comment (carlosgcampos) [inline Source/JavaScriptCore/API/glib/JSCDefines.h.in]: Could we include the class definition here too?

File: Source/JavaScriptCore/GLib.cmake [hunk 5]
```diff
@@ -27,21 +30,30 @@ list(APPEND JavaScriptCore_INTERFACE_INCLUDE_DIRECTORIES
...
 
     ${JavaScriptCoreGLib_DERIVED_SOURCES_DIR}/jsc/JSCVersion.h
 )
 
+GENERATE_GLIB_API_HEADERS(JavaScriptCore JavaScriptCore_HEADER_TEMPLATES
+    ${JavaScriptCoreGLib_DERIVED_SOURCES_DIR}/jsc
+    JavaScriptCore_INSTALLED_HEADERS
+    "-DENABLE_2022_GLIB_API=$<BOOL:${ENABLE_2022_GLIB_API}>"
+)
+
```
Comment (carlosgcampos) [inline Source/JavaScriptCore/GLib.cmake]: Could we split the patch and do the jsc in a different patch? it changes more things like autogenerating the headers.

**Review by carlosgcampos — COMMENTED** [review]
I like this approach. Thanks!

File: Source/JavaScriptCore/GLib.cmake (hunk 5 above)
Comment (aperezdc) [reply Source/JavaScriptCore/GLib.cmake]: Sure, I can split the JSC API changes in a separate patch.

File: Source/JavaScriptCore/API/glib/JSCContextInternal.h (hunk 1 above)
Comment (aperezdc) [reply Source/JavaScriptCore/API/glib/JSCContextInternal.h]: Now that I think of it with a fresher mind, probably there was a `${JavaScriptCoreGLib_DERIVED_SOURCES}/jsc" entry missing in the include paths.

I will be splitting the JSC changes into a separate PR so I will look into this later.

File: Source/JavaScriptCore/API/glib/JSCContextPrivate.h (hunk 2 above)
Comment (aperezdc) [reply Source/JavaScriptCore/API/glib/JSCContextPrivate.h]: This was a sleepy fix for the same missing include path as above.

File: Source/JavaScriptCore/API/glib/JSCDefines.h.in (hunk 3 above)
Comment (aperezdc) [reply Source/JavaScriptCore/API/glib/JSCDefines.h.in]: We need the declarations generated by the `G_DECLARE_FINAL_TYPE` macro to be part of the public headers, and I tried to move the instance struct to the implementation side but I hit some roadblock so I thought of doing it in a follow-up patch. I can give it a try again today after, maybe after sleeping the night I will manage to get it done 😉️

File: Source/JavaScriptCore/API/glib/JSCDefines.h.in (hunk 4 above)
Comment (aperezdc) [reply Source/JavaScriptCore/API/glib/JSCDefines.h.in]: I'll try in the patch I will be splitting for the JSC changes 🫡️

File: Source/JavaScriptCore/API/glib/JSCDefines.h.in (hunk 3 above)
Comment (mcatanzaro) [reply Source/JavaScriptCore/API/glib/JSCDefines.h.in]: Should definitely try to move the instance structs of final types to the cpp.

**Comment by mcatanzaro** [comment]
Note typo in commit message: "used by the for the"

File: Source/JavaScriptCore/API/glib/JSCClass.h.in [hunk 6]
```diff
@@ -31,11 +31,12 @@
...
 
 #define JSC_TYPE_CLASS            (jsc_class_get_type())
+#if !ENABLE(2022_GLIB_API)
 #define JSC_CLASS(obj)            (G_TYPE_CHECK_INSTANCE_CAST((obj), JSC_TYPE_CLASS, JSCClass))
 #define JSC_IS_CLASS(obj)         (G_TYPE_CHECK_INSTANCE_TYPE((obj), JSC_TYPE_CLASS))
+#endif
 
-typedef struct _JSCClass JSCClass;
-typedef struct _JSCClassClass JSCClassClass;
+JSC_DECLARE_FINAL_TYPE(JSCClass, jsc_class, JSC, CLASS, GObject)
```
Comment (mcatanzaro) [inline Source/JavaScriptCore/API/glib/JSCClass.h.in]: Um, you might hate me for this suggestion, but since this is a public header and not the .cpp, there should be a space before the opening paren:

```suggestion
JSC_DECLARE_FINAL_TYPE (JSCClass, jsc_class, JSC, CLASS, GObject)
```

Yeah... in all the public headers....

File: Source/JavaScriptCore/API/glib/JSCDefines.h.in (hunk 3 above)
Comment (mcatanzaro) [reply Source/JavaScriptCore/API/glib/JSCDefines.h.in]: (only in the new API version, of course)

File: Source/JavaScriptCore/API/glib/JSCDefines.h.in (hunk 4 above)
Comment (mcatanzaro) [reply Source/JavaScriptCore/API/glib/JSCDefines.h.in]: No? That would only work if every class struct is identical?

File: Source/JavaScriptCore/API/glib/JSCDefines.h.in (hunk 3 above)
Comment (mcatanzaro) [reply Source/JavaScriptCore/API/glib/JSCDefines.h.in]: Yeah so this definitely does not belong here because it prevents us from ever getting rid of `priv` during the lifetime of the new API. There are two better ways to do it:

 * Open code it in each .cpp file inside `#if ENABLE(2022_GLIB_API)` guards.
 * Have `WEBKIT_DEFINE_FINAL_TYPE` do it. I think this is probably nicest.

File: Source/JavaScriptCore/API/glib/JSCClass.h.in (hunk 6 above)
Comment (aperezdc) [reply Source/JavaScriptCore/API/glib/JSCClass.h.in]: 🤪️

File: Source/WebKit/UIProcess/API/glib/WebKitDefines.h.in [hunk 7]
```diff
@@ -50,4 +50,71 @@
...
+    return G_TYPE_CHECK_CLASS_CAST (ptr, module_obj_name##_get_type (), ModuleObjName##Class); }         \
+  G_GNUC_UNUSED static inline gboolean MODULE##_IS_##OBJ_NAME (gpointer ptr) {                           \
+    return G_TYPE_CHECK_INSTANCE_TYPE (ptr, module_obj_name##_get_type ()); }                            \
+  G_GNUC_UNUSED static inline gboolean MODULE##_IS_##OBJ_NAME##_CLASS (gpointer ptr) {                   \
+    return G_TYPE_CHECK_CLASS_TYPE (ptr, module_obj_name##_get_type ()); }                               \
+  G_GNUC_UNUSED static inline ModuleObjName##Class * MODULE##_##OBJ_NAME##_GET_CLASS (gpointer ptr) {    \
+    return G_TYPE_INSTANCE_GET_CLASS (ptr, module_obj_name##_get_type (), ModuleObjName##Class); }       \
+  G_GNUC_END_IGNORE_DEPRECATIONS
+
+#define WEBKIT_DECLARE_DERIVABLE_TYPE WEBKIT_DECLARE_TYPE
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/API/glib/WebKitDefines.h.in]: Nit: the `WEBKIT_DECLARE_TYPE` is unneeded because it's only used here. Can simply rename it to `WEBKIT_DECLARE_DERIVABLE_TYPE`.

BTW, I agree this is the best-possible solution.

File: Source/JavaScriptCore/API/glib/JSCClass.h.in (hunk 6 above)
Comment (aperezdc) [reply Source/JavaScriptCore/API/glib/JSCClass.h.in]: (Nothing that `sed` wouldn't do automatically, BTW)

File: Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.cpp [hunk 8]
```diff
@@ -72,15 +72,7 @@ struct _WebKitWebHitTestResultPrivate {
...
 
-#if ENABLE(2022_GLIB_API)
-struct _WebKitWebHitTestResultClass {
-    GObjectClass parent;
-};
-
 WEBKIT_DEFINE_FINAL_TYPE(WebKitWebHitTestResult, webkit_web_hit_test_result, G_TYPE_OBJECT)
-#else
-WEBKIT_DEFINE_TYPE(WebKitWebHitTestResult, webkit_web_hit_test_result, WEBKIT_TYPE_HIT_TEST_RESULT)
-#endif
```
Comment (mcatanzaro) [inline Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.cpp]: Oops, careful here! You broke WebKitWebHitTestResult in the pre-2022 API.

File: Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebExtensionAutocleanups.h.in [hunk 9]
```diff
@@ -25,9 +25,16 @@
...
+ */
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (WebKitUserMessage, g_object_unref)
+#endif
+
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (WebKitFrame, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (WebKitScriptWorld, g_object_unref)
-G_DEFINE_AUTOPTR_CLEANUP_FUNC (WebKitUserMessage, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (WebKitWebEditor, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (WebKitWebExtension, g_object_unref)
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (WebKitWebHitTestResult, g_object_unref)
```
Comment (mcatanzaro) [inline Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebExtensionAutocleanups.h.in]: It looks like you intentionally did not convert these? Why?

And you did convert WebKitWebHitTestResult, though...?

File: Source/JavaScriptCore/API/glib/JSCDefines.h.in (hunk 3 above)
Comment (aperezdc) [reply Source/JavaScriptCore/API/glib/JSCDefines.h.in]: My plan was to try piggybacking this into `WEBKIT_DEFINE_FINAL_TYPE` 🛠️; it's just taking me some time&mdash;please bear with me because every set of changes involves making at least three different builds (WPE, GTK3, GTK4) to make sure things go in the right direction.

File: Source/JavaScriptCore/API/glib/JSCDefines.h.in (hunk 4 above)
Comment (aperezdc) [reply Source/JavaScriptCore/API/glib/JSCDefines.h.in]: In the case of JSC API classes, none has methods in the vtable, so we can generate them all from the macros. Anyway, that will be in the separate follow-up with the JSC part.

File: Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.cpp (hunk 8 above)
Comment (aperezdc) [reply Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.cpp]: Good catch, thanks! I was a bit trigger-happy removing things around 😉

File: Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebExtensionAutocleanups.h.in (hunk 9 above)
Comment (aperezdc) [reply Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebExtensionAutocleanups.h.in]: I've converted now all the types from the web extensions API.

File: Source/WebKit/UIProcess/API/glib/WebKitDefines.h.in (hunk 7 above)
Comment (aperezdc) [reply Source/WebKit/UIProcess/API/glib/WebKitDefines.h.in]: The `WEBKIT_DECLARE_TYPE` macro is used in `WebKitDownload.h.in`, `WebKitWebContext.h.in`, and `WebKitWebHitTestResult.h.in`. While right now it's the same as `WEBKIT_DECLARE_DERIVABLE_TYPE`, I would find it confusing that we use something named “derivable” to define a final type like `WebKitWebContext` or `WebKitDownload`, and at the same time we cannot use `WEBKIT_DECLARE_FINAL_TYPE` there because it hardcodes the class struct to contain nothing but these types do use a custom class struct with vtable functions.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitDefines.h.in]: OK... but the custom class struct does not belong in the modern API version because the type is no longer derivable and so there's no point to having vfuncs because they can never be used. So I would use `WEBKIT_DECLARE_FINAL_TYPE` anyway and move the class struct inside `#if !ENABLE(2022_GLIB_API)`.

Beware that does require some small changes in the implementation file, e.g. for WebKitDownload:

```
diff --git a/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp b/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
index 44367ee66541..b3459268c1cd 100644
--- a/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
+++ b/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
@@ -283,7 +283,11 @@ static void webkit_download_class_init(WebKitDownloadClass* downloadClass)
         "decide-destination",
         G_TYPE_FROM_CLASS(objectClass),
         G_SIGNAL_RUN_LAST,
+#if ENABLE(2022_GLIB_API)
+        0,
+#else
         G_STRUCT_OFFSET(WebKitDownloadClass, decide_destination),
+#endif
         g_signal_accumulator_true_handled, NULL,
         g_cclosure_marshal_generic,
         G_TYPE_BOOLEAN, 1,
```

Comment (aperezdc) [reply Source/WebKit/UIProcess/API/glib/WebKitDefines.h.in]: @mcatanzaro It's not that easy, the above diff is not enough, what you want is this:

```diff
diff --git a/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp b/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
index 586a01a90ea9..c7c4cdf52c4e 100644
--- a/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
+++ b/Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
@@ -150,7 +150,9 @@ static void webkit_download_class_init(WebKitDownloadClass* downloadClass)
     objectClass->set_property = webkitDownloadSetProperty;
     objectClass->get_property = webkitDownloadGetProperty;
 
+#if !ENABLE(2022_GLIB_API)
     downloadClass->decide_destination = webkitDownloadDecideDestination;
+#endif
 
     /**
      * WebKitDownload:destination:
@@ -283,7 +285,11 @@ static void webkit_download_class_init(WebKitDownloadClass* downloadClass)
         "decide-destination",
         G_TYPE_FROM_CLASS(objectClass),
         G_SIGNAL_RUN_LAST,
-        G_STRUCT_OFFSET(WebKitDownloadClass, decide_destination),
+#if ENABLE(2022_GLIB_API)
+        0,
+#else
+        G_STRUCT_OFFSET(WebKitDownloadClass,  decide_destination),
+#endif
         g_signal_accumulator_true_handled, NULL,
         g_cclosure_marshal_generic,
         G_TYPE_BOOLEAN, 1,
@@ -408,8 +414,16 @@ String webkitDownloadDecideDestinationWithSuggestedFilename(WebKitDownload* down
 {
     if (download->priv->isCancelled)
         return emptyString();
-    gboolean returnValue;
-    g_signal_emit(download, signals[DECIDE_DESTINATION], 0, suggestedFilename.data(), &returnValue);
+
+    gboolean handled = FALSE;
+    g_signal_emit(download, signals[DECIDE_DESTINATION], 0, suggestedFilename.data(), &handled);
+
+#if ENABLE(2022_GLIB_API)
+    if (!handled)
+        handled = webkitDownloadDecideDestination(download, suggestedFilename.data());
+#endif
+    ASSERT(handled);
+
     allowOverwrite = download->priv->allowOverwrite;
     GUniquePtr<char> destinationPath(g_filename_from_uri(download->priv->destinationURI.data(), nullptr, nullptr));
     if (!destinationPath)
```

While not too complicated, and in most cases (if not all) the other types which currently have a public class struct with vfuncs can be fixed in similar ways&mdash;but this patch is already big enough, so I would rather do these changes in a follow-up patch. WDYT?

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitDefines.h.in]: Sure, of course a follow-up patch is fine.

Comment (aperezdc) [reply Source/WebKit/UIProcess/API/glib/WebKitDefines.h.in]: So... r+ for this one, then? I think all the other issues have been tackled 😄

**Review by mcatanzaro — APPROVED** [review]
